### PR TITLE
Automatically requeue jobs when responses can't be delivered

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
@iwiznia @flodnv @mea36 

This is a relatively small change that adds an interface for a plugin to handle a command when the response from running that command couldn't be delivered to the client, and has `Jobs` implement that to re-queue the results so they can be run again. This means that when bedrock is slow/behind or otherwise is disconnected from a webserver before being able to send a reply to a `GetJob` or `GetJobs` command, those jobs are not "stuck".

This is meant to address this issue:
https://github.com/Expensify/Expensify/issues/85100

And prevent us having to do these sorts of tasks in the future:
https://github.com/Expensify/Expensify/issues/90281

Granted, this is not the only possible way to create a stuck job, but this seems to be the biggest single cause of stuck jobs for us currently.

### CODE REVIEW

File: BedrockServer.cpp [hunk 1]
```diff
@@ -1503,6 +1509,13 @@ void BedrockServer::_reply(BedrockCommand& command) {
     else if (!SIEquals(command.request["Connection"], "forget")) {
         SINFO("No socket to reply for: '" << command.request.methodLine << "' #" << command.initiatingClientID);
     }
+
+    // If we didn't actually send the response, let the plugin know about that.
+    if (!replySent) {
+        if (command.processedBy) {
```
Comment (iwiznia) [inline BedrockServer.cpp]: You can combine these 2 ifs into 1

File: plugins/Jobs.cpp [hunk 2]
```diff
@@ -1195,6 +1196,34 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
         return true;
     }
 
+    // Requeue a job for which a getJob(s) command could not complete.
+    else if (SIEquals(requestVerb, "RequeueJobs")) {
```
Comment (iwiznia) [inline plugins/Jobs.cpp]: By doing this, you are exposing this new command, right? Should we try to mark this as internal somehow or not worth it?

File: plugins/Jobs.cpp [hunk 3]
```diff
@@ -1195,6 +1196,34 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
+        string selectQuery = "SELECT jobID, retryAfter FROM jobs WHERE jobID IN(" + SQList(jobIDs)+ ");";
+        db.read(selectQuery, result);
+
+        for (auto& row : result.rows) {
+            int64_t jobID = stoll(row[0]);
+            bool retryAfter = !row[1].empty();
+            string updateQuery;
+            if (retryAfter) {
+                // Reset the nextRun time to now, we don't need to wait for the timeout.
+                updateQuery = "UPDATE jobs SET nextRun = DATETIME("+ SCURRENT_TIMESTAMP() + ") WHERE jobID = " + SQ(jobID) + ";";
```
Comment (iwiznia) [inline plugins/Jobs.cpp]: We do this so it remains in `RUNQUEUED` state instead of `RUNNING` right?

File: plugins/Jobs.cpp [hunk 4]
```diff
@@ -1290,3 +1319,41 @@ bool BedrockPlugin_Jobs::_isValidSQLiteDateModifier(const string& modifier) {
...
+            STable jobsJSON = SParseJSONObject(command.response.content);
+            list<string> jobs = SParseJSONArray(jobsJSON["jobs"]);
+            for (auto& job : jobs) {
+                STable jobJSON = SParseJSONObject(job);
+                if (jobJSON.find("jobID") != jobJSON.end()) {
+                    jobIDs.push_back(jobJSON["jobID"]);
+                }
+            }
+        }
+        SWARN("Failed sending response to '" << command.request.methodLine << "', re-queueing jobs: "<< SComposeList(jobIDs));
```
Comment (iwiznia) [inline plugins/Jobs.cpp]: Do we want to warn here? Given we know is "normal" and there's nothing else we can do, we should log hmmm or info.

File: plugins/Jobs.cpp [hunk 5]
```diff
@@ -1290,3 +1319,41 @@ bool BedrockPlugin_Jobs::_isValidSQLiteDateModifier(const string& modifier) {
...
+        if(_server) {
```
Comment (iwiznia) [inline plugins/Jobs.cpp]: Can this be false? If so, what would happen with the jobs? Maybe we should alert in the else saying we could not revert the changes?

File: plugins/Jobs.cpp [hunk 6]
```diff
@@ -1290,3 +1319,41 @@ bool BedrockPlugin_Jobs::_isValidSQLiteDateModifier(const string& modifier) {
...
+            SData requeue("RequeueJobs");
+            requeue["jobIDs"] = SComposeList(jobIDs);
+
+            // Keep the request ID so we'll be able to associate these in the logs.
```
Comment (iwiznia) [inline plugins/Jobs.cpp]: Nice!

File: test/tests/jobs/FailedJobReplyTest.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    void tearDown() {
+        SData command("Query");
+        command["query"] = "DELETE FROM jobs WHERE jobID > 0;";
+        tester->executeWaitVerifyContent(command);
+    }
+
+    void tearDownClass() { delete tester; }
+
+    // Cannot cancel a job with children
+    void failSendingResponse() {
```
Comment (iwiznia) [inline test/tests/jobs/FailedJobReplyTest.cpp]: I think we should add a test for jobs with retryAfter too.

File: plugins/Jobs.cpp [hunk 8]
```diff
@@ -1195,6 +1196,34 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
...
+
+        SQResult result;
+        string selectQuery = "SELECT jobID, retryAfter FROM jobs WHERE jobID IN(" + SQList(jobIDs)+ ");";
+        db.read(selectQuery, result);
+
+        for (auto& row : result.rows) {
+            int64_t jobID = stoll(row[0]);
+            bool retryAfter = !row[1].empty();
+            string updateQuery;
+            if (retryAfter) {
```
Comment (mea36) [inline plugins/Jobs.cpp]: Why does it matter if the the job has a `retryAfter` or not? 
I think all the jobs should be updated to `state='QUEUED'` and `nextRun=DATETIME('now())` since they were about to be run anyway

The tricky cases are
1.  `retryAfter` is not empty, but this is actually irrelevant because the job would of been retried again anyway so we might it well retry it now
1.  `repeat` is not empty.  This is a bit trickier.  At first I was thinking we should respect the `repeat` value and therefore was considering proposing that we use `RetryJob` instead of `RequeueJobs` since `RetryJob` handles all the weird cases already. But `repeat` values can be set up so the job runs once a year and it seems silly to wait a year to run a job because the server timed out.

File: plugins/Jobs.cpp (hunk 2 above)
Comment (tylerkaraszewski) [reply plugins/Jobs.cpp]: It's only exposed to the same places that can make new jobs anyway, I don't see how there's any greater risk here.

File: plugins/Jobs.cpp (hunk 3 above)
Comment (tylerkaraszewski) [reply plugins/Jobs.cpp]: `RUNQUEUED` jobs normally stay `RUNQUEUED` until they're finished. We could do nothing here and they'd eventually retry anyway. We do this so they can retry sooner. If they have an 8-hour timeout, and we do nothing in this case, they'll sit for 8 hours, but since we already know they're going to need to retry, we might as well let them retry right away. Like @mea36 says, we could also make them `QUEUED` again, it doesn't really have any different effect.

File: plugins/Jobs.cpp (hunk 8 above)
Comment (tylerkaraszewski) [reply plugins/Jobs.cpp]: Ok, I think just always making them `QUEUED` with a `nextRun` time of right now is simpler and has the same effect.

As far as `repeat` goes, I think I agree with your conclusion - these jobs weren't run actually run and thus this shouldn't affect when they repeat, we should try to run them again as soon as possible, and then they will repeat as normal.

File: test/tests/jobs/FailedJobReplyTest.cpp (hunk 7 above)
Comment (tylerkaraszewski) [reply test/tests/jobs/FailedJobReplyTest.cpp]: Yeah, there are several more tests to add.

**Comment by quinthar** [comment]
> before being able to send a reply to a GetJob or GetJobs command, those jobs are not "stuck".
To clarify, how do we know we didn't send it?  Does this only happen if we haven't even sent it from our send queue, or merely that we haven't been able to *confirm* it was sent before the connection died?  I think we want to err on the side of caution and if there is *any* chance that we in fact sent it, assume we did.

File: test/tests/jobs/FailedJobReplyTest.cpp [hunk 9]
```diff
@@ -23,27 +23,110 @@ struct FailedJobReplyTest : tpunit::TestFixture {
...
+                    STable j;
+                    for (auto& nvp : job) {
+                        j[nvp.first] = nvp.second;
+                    }
+                    if (retryAfter) {
+                        j["retryAfter"] = "+1 HOUR";
+                    }
+                    jobs.push_back(SComposeJSONObject(j));
+                }
+                STable body;
```
Comment (mea36) [inline test/tests/jobs/FailedJobReplyTest.cpp]: This isn't used anywhere
